addons/account/models/account_move.py

```diff
@@ -1346,6 +1352,122 @@ def _compute_payments_widget_reconciled_info(self):
+            ))
+        return epd_lines
+
+    def _prepare_cash_rounding_base_line_for_taxes_computation(self, cash_rounding_line):
+        self.ensure_one()
+        sign = self.direction_sign
+        rate = self.invoice_currency_rate
+
+        return self.env['account.tax']._prepare_base_line_for_taxes_computation(
+            cash_rounding_line,
+            price_unit=sign * cash_rounding_line.amount_currency,
+            quantity=1.0,
+            sign=sign,
+            special_mode='total_excluded',
+            special_type='cash_rounding',
+
+            is_refund=self.move_type in ('out_refund', 'in_refund'),
+            rate=rate,
+        )
+
+    def _prepare_tax_line_for_taxes_computation(self, tax_line):
+        self.ensure_one()
+        return self.env['account.tax']._prepare_tax_line_for_taxes_computation(
+            tax_line,
+            sign=self.direction_sign,
+        )
+
+    def _get_rounded_base_lines(self, round_from_tax_lines=True):
+        self.ensure_one()
+        AccountTax = self.env['account.tax']
+        is_invoice = self.is_invoice(include_receipts=True)
+        if self.id or not is_invoice:
+            base_amls = self.line_ids.filtered(lambda line: line.display_type == 'product')
+        else:
+            base_amls = self.invoice_line_ids.filtered(lambda line: line.display_type == 'product')
+        base_lines = [self._prepare_product_base_line_for_taxes_computation(line) for line in base_amls]
+
+        tax_lines = []
+        if self.id:
+            # The invoice is stored so we can add the early payment discount lines directly to reduce the
+            # tax amount without touching the untaxed amount.
+            epd_amls = self.line_ids.filtered(lambda line: line.display_type == 'epd')
+            base_lines += [self._prepare_epd_base_line_for_taxes_computation(line) for line in epd_amls]
+            cash_rounding_amls = self.line_ids \
+                .filtered(lambda line: line.display_type == 'rounding' and not line.tax_repartition_line_id)
+            base_lines += [self._prepare_cash_rounding_base_line_for_taxes_computation(line) for line in cash_rounding_amls]
+            AccountTax._add_base_lines_tax_details(base_lines, self.company_id)
+            tax_amls = self.line_ids.filtered('tax_repartition_line_id')
+            tax_lines = [self._prepare_tax_line_for_taxes_computation(tax_line) for tax_line in tax_amls]
+            AccountTax._round_base_lines_tax_details(base_lines, self.company_id, tax_lines=tax_lines if round_from_tax_lines else [])
+        else:
+            # The invoice is not stored yet so the only thing we have is the invoice lines.
+            base_lines += self._prepare_epd_base_lines_for_taxes_computation_from_base_lines(base_amls)
+            AccountTax._add_base_lines_tax_details(base_lines, self.company_id)
+            AccountTax._round_base_lines_tax_details(base_lines, self.company_id)
+        return base_lines, tax_lines
```

Comment:
I don't see where the grouping is made, to have several invoice lines with same tax considered as a single base... au secours

Reply:
All the generation of tax lines is done in `_sync_tax_lines`.